File: 11_动态规划/力扣加加/区间dp/664. 奇怪的打印机.ts
/**
 * @param {string} s
 * @return {number}
 * 打印机每次只能打印由 同一个字符 组成的序列
 * 每次可以在任意起始和结束位置打印新字符，并且会覆盖掉原来已有的字符
 * https://leetcode-cn.com/problems/strange-printer/solution/yi-wen-tuan-mie-qu-jian-dp-by-bnrzzvnepe-24jz/
 */
const strangePrinter = function (s: string): number {
  if (!s) return 0
  const n = s.length

  const dp = Array.from({ length: n + 1 }, () => Array(n).fill(0))
  for (let i = 0; i < n; i++) {
    dp[i][i] = 1
  }
  // ：从前向后
  // 区间长度 区间起点 区间终点 初始化 枚举分割点 首位一样可减少一次
  for (let l = 1; l < n; l++) {
    for (let i = 0; i < n - l; i++) {
      const j = i + l
      // dp[i][j]的上限：比前面多一次
      dp[i][j] = dp[i + 1][j] + 1
      for (let k = i + 1; k <= j; k++) {
        // 与首部相同 少打一次
        if (s[k] === s[i]) dp[i][j] = Math.min(dp[i][j], dp[i][k - 1] + dp[k + 1][j])
      }
    }
  }
  console.table(dp)
  return dp[0][dp[0].length - 1]
}

console.log(strangePrinter('aaabbb'))
// 输出：2
// 解释：首先打印 "aaa" 然后打印 "bbb"。
console.log(strangePrinter('aba'))
// 输出：2
// 解释：首先打印 "aaa" 然后在第二个位置打印 "b" 覆盖掉原来的字符 'a'。

// 与 312思想基本一致，枚举 k 时加一层限制
// 这道题每次可以在任意起始和结束位置打印新字符。 因此我们需要暴力枚举所有的起始位置和结束位置的笛卡尔积

File: 11_动态规划/dp分类/区间dp/染色/664. 奇怪的打印机.py
# 664. 奇怪的打印机
# https://leetcode.cn/problems/strange-printer/description/
#
# 有台奇怪的打印机有以下两个特殊要求：
#
# 打印机每次只能打印由 同一个字符 组成的序列。
# 每次可以在任意起始和结束位置打印新字符，并且会覆盖掉原来已有的字符。
# 给你一个字符串 s ，你的任务是计算这个打印机打印它需要的最少打印次数。
# 1 <= s.length <= 100

from functools import lru_cache
from itertools import groupby


class Solution:
    def strangePrinter(self, s: str) -> int:
        values = [a for a, _ in groupby(s)]

        @lru_cache(None)
        def dfs(start: int, end: int) -> int:
            if start >= end:
                return 0

            res = dfs(start + 1, end) + 1  # !直接移除这一段
            for mid in range(start + 1, end):
                if values[mid] == values[start]:
                    res = min(
                        res, dfs(start + 1, mid) + dfs(mid, end)
                    )  # !处理这一段后，start和mid合并
            return res

        return dfs(0, len(values))


print(Solution().strangePrinter("aaabbb"))
# 输出：2
# 解释：首先打印 "aaa" 然后打印 "bbb"。


# CABBA
#  CA | BB | A
# It was simply inserted with the cost of 1
# It was free from some previous step to the left that printed this character already (we can print extra character all the way till the end)
# 打一个新的要1快钱  cost = dfs(s[:-1]) + 1
# 打一个前面相同的不要钱   dfs(s[: i + 1]) + dfs(s[i + 1 : -1])
